### ISSUE TITLE
Enhance wins sheets to support selection for homepage randomizer

### ISSUE THREAD

**Issue by macho-catt** [hackforla/website#2505]
### Dependency
- [x] #2901 
- [x] #4134
- [x] #3641

### Overview
We need enhancements to Wins-form (Responses) to help PMs in identifying which wins entries should be displayed in the website homepage randomizer.

### Action Items
- [ ] Follow sections 1-3 in the Wins Apps Script Development Process[^4].  Note that in this issue you will only be modifying the Apps Script `wins-form-responses` so all commands should be run from `/google-apps-script/wins-form-responses` 
- [ ] As described in section 3a and 3b, you will provide a dev lead with the google account you wish to use for testing, and request the dev lead to:
   - [ ] share the Wins Form Admin Guide[^1] as VIEWER 
   - [ ] share the Wins-form (Response)[^2] sheet as VIEWER
   - [ ] share the Wins-form[^3] 
      - [ ] Note to lead: make a copy of `Wins-form`, including the issue # in the name, then share that copy with the developer as EDITOR.  
- [ ] Continue following instructions in 3c and d.  As described in 3d, you will have to make modifications to certain documents in the test folder.  
   - [ ] Update settings in Wins-form so that responses are stored in your test version of Wins-form (Response)
   - [ ] In Wins-form (Response), add trigger so that `insertLatestFormSubmitIntoReviewSheet` runs when the test form is submitted.  
- [ ] Continue following instructions through section 6 "Editing".  As described in this section, editing can be done locally in an IDE or in the Google Drive cloud editor, and the `clasp` utility can be used to transfer code back and forth between the local worksation and the Google Drive as needed.   
- [ ] Copy columns H through R from  the "Homepage Summary" sheet to the "Review" sheet.  
- [ ] Create a new sheet "Log".  Label the fist column "Wins ID" and copy columns F through H from the "Homepage Summary" sheet to the "Log" sheet, to the right of the "Wins ID" column
- [ ] Setup a formula so that column "# of times appeared on website" on the Review sheet, is sourced from the "Log" sheet, based on matching the "Wins ID" columns on the "Review" and "Log" sheets.  
- [ ] On the Review sheet, add a new column "Person ID" to the right of "Wins ID"
- [ ] Create a new tab/sheet "Persons" with columns "Person ID", "email", "# of wins IDs"
- [ ] Add code so that when a new wins submission is processed, a "Person ID" is assigned, and the "# of wins IDs" is populated, based on the assumption that each unique email address represents a unique "person".  
- [ ] Test the code to confirm that both for new form submissions and for Review sheet updates, both the Review, Log, and Persons sheets are updated appropriately.  
- [ ] You will not have to run `main()` for this issue, so if there are errors related to not having access to gh-requests, you can remove the gh-requests library and if necessary "comment out" main()
- [ ] Follow all steps through section 8 "Completing an issue".  For this issue, only a single file `google-apps-script/wins-form-responses/Code.js` should be committed.  Contact the dev lead via Slack to arrange the review.  

### Merge Team/dev Lead
- [ ] ensure that all access to Wins Form Admin Guide[^1], the Wins-form (Response) Sheet[^2], and the Wins-form[^3] have been revoked.  

### Resources/Instructions

- [Google Apps Script](https://developers.google.com/apps-script/overview)
- [_wins-data file](https://github.com/hackforla/website/blob/gh-pages/_data/external/_wins-data.json)
- [wins page JS](https://github.com/hackforla/website/blob/gh-pages/assets/js/wins.js)
[^1]:[Wins Form Admin Guide](https://docs.google.com/document/d/1j1JTX5XzotgVCVZ91ImvenedPN5-IVjdSKDukq1B9iE/edit?usp=drive_link)
[^2]:[Wins-form (Response) Sheet](https://docs.google.com/spreadsheets/d/1fXmYrmNtrgdzkM_odGIRbSwdaea3yS0cTLCJUzZ6Sc0/edit?usp=drive_link)
[^3]:[Wins-form](https://docs.google.com/forms/d/1wojt_jriqXe4388wpIqC-b3dzX2he-y1oa4lbbLCRvI/edit)
[^4]:[Wins Apps Script Development Process](https://github.com/hackforla/website/issues/4134#issuecomment-1741851926)

**Comment by SAUMILDHANKAR** [comment]
Availability: 10 hrs per week. Tuesday and Thursday evenings.
ETA: 1/10/22

**Comment by SAUMILDHANKAR** [comment]
Progress: Went over different sections of code.gs file. Will start creating the new section next and test adding data from review sheet to response sheet in personal copy and repo.
Blockers: Need more analysis for the follwing column headers:
Date Live on homepage start | Date Live on homepage end | # of times appeared on website
Availability: 5 hrs
ETA: 1/15/22

**Comment by SAUMILDHANKAR** [comment]
Progress: Didn't make much progress, since took a lot of days off last week.
Blockers: Need more analysis for the following column headers:
Date Live on homepage start | Date Live on homepage end | # of times appeared on website
Availability: 5 hrs
ETA: 1/30/22

**Comment by SAUMILDHANKAR** [comment]
Progress: Added a row on the homepage summary sheet with formulas to calculate respective number of positions (eg. engineers, designer, data analyst and so on) which have TRUE on the Homepage.
Blockers: Will need help to access bot account for testing.
Availability: 3 hrs
ETA: 2/8/22

**Comment by SAUMILDHANKAR** [comment]
Progress: Identified an error in the response sheet, this might be related to the function insertLatestFormSubmitIntoReviewSheet. Will work on resolving this as similar function might be required for this issue.
Blockers: No blockers right now.
Availability: 3 hrs
ETA: 2/15/22

**Comment by SAUMILDHANKAR** [comment]
Progress: Didn't work on this issue last week. Made progress on the issue to fix the bug in 2 weeks inactive label. 
Blockers: No blockers right now.
Availability: 3 hrs
ETA: 2/25/22

**Comment by SAUMILDHANKAR** [comment]
Progress: Created a new function that adds Google form data to the summary sheet, added an on form submit trigger to the function. Almost complete, need to test on the final version.
Blockers: Final testing to be done once #2901 is merged.
Availability: Not aplicable
ETA: To be determined.

**Comment by SAUMILDHANKAR** [comment]
Moving this issue to the ice box column and adding dependency since demo to the team won't work until #2901 is fixed.
<details><summary>Code so far that adds data to the Summary1 sheet</summary>

```
/**
/************************************************** TRIGGER("On Form Submit") 4 SECTION ********************************************************************/
/*
  For each new wins submission, add data to the "Summary1" Sheet.
*/

function insertLatestFormSubmitIntoSummarySheet(formSubmitEvent){

	const rawFormDataObject = formSubmitEvent.namedValues;
	const formDataInsertedAt = formSubmitEvent.range; 
  
	const reviewSheet = SpreadsheetApp.getActiveSpreadsheet().getSheetByName("Review1");
	const responsesSheet = SpreadsheetApp.getActiveSpreadsheet().getSheetByName("Responses");
	const summarySheet = SpreadsheetApp.getActiveSpreadsheet().getSheetByName("Summary1");

	let processedFormDataObject = new Map();
	processedFormDataObject.set("Email", rawFormDataObject["Email Address"])
	processedFormDataObject.set("Name", rawFormDataObject["Full name"])
	processedFormDataObject.set("LinkedIn", rawFormDataObject["Linkedin URL (optional)"])
	processedFormDataObject.set("LinkedIn Picture Allowed ?", rawFormDataObject["Could we use your Linkedin profile picture next to your story?"])
	processedFormDataObject.set("GitHub", rawFormDataObject["Github URL (optional)"])
	processedFormDataObject.set("GitHub Picture Allowed ?", rawFormDataObject["Could we use your Github profile picture next to your story?"])
	processedFormDataObject.set("Teams", rawFormDataObject["Select the team(s) you're on"])
	processedFormDataObject.set("Roles", rawFormDataObject["Select your role(s) on the team"])
	processedFormDataObject.set("Specific Roles", rawFormDataObject["What is/was your specific role? (optional)"])
	processedFormDataObject.set("Celebrations", rawFormDataObject["What do you want to celebrate (select all that apply)?"])
	processedFormDataObject.set("Overview", rawFormDataObject["Give us a brief overview"])

	const memberSince = rawFormDataObject["When did you join Hack for LA? (optional)"]

	// Prepare display string from proccssed form data object
	let displayString = "";
	for(const [key,value] of processedFormDataObject ){
        displayString += `${key} : ${value} \n\n`;
	}

	// Create and insert a new empty row before row 2
	summarySheet.insertRowBefore(4);
	
	// Insert member since value on column 3
	summarySheet.getRange(4,3).setValue(memberSince);
	
	// Insert display overview value on column 4
	summarySheet.getRange(4,4).setValue( displayString );
	
	// Set the column 'A'  for new row created in the 'Summary1' sheet by form submission to be mapped to the new row created in the 'review1' sheet
	summarySheet.getRange(4,1).setFormula("=Review1!A3");
	
	// Set the column 'B'  for new row created in the 'Summary1' sheet by form submission to be mapped to the new row created in the 'review1' sheet
	summarySheet.getRange(4,2).setFormula("=Review1!B3");
	
	// Set the column 'E'  for new row created in the 'Summary1' sheet by form submission to be mapped to the new row created in the 'review1' sheet
	summarySheet.getRange(4,5).setFormula("=Review1!E3");
	
	// Set formula to count the role of each row
	summarySheet.getRange(4,9).setFormula("=COUNTIF(E4,$I$1)");
	
}  
```

</details>

**Comment by roslynwythe** [comment]
Availability: 12/18 10 am - 8 pm; 12/19 9 am - noon; 12/20 5 pm - 10 pm; 10/21 10 am - 5 pm
ETA: 12/30

**Comment by roslynwythe** [comment]
Hi @arpitapandya @macho-catt - The issue description, states that I need to make a copy of the Wins form response sheet and Secret files in my own drive.  I have the copy of the Wins form response sheet but I need permission/advice how to copy the Secret files,  I'm not certain but I don't think I have access to the Hfla website admin folder;  I couldn't find the link to that folder.  I was surprised that the issue says that I should get access to the Secret files, since the Wins Form Admin Guide says that only merge-team and admin accounts should have access, and for this issue I don't need to see the super-sensitive key files directly; all I need is to be able to call the gh-library from my account to test my code changes to the Response sheet script, but presumably that is not possible without the Secret file and container.  

I believe that for testing I need a copy of the Google form; is it possible to share a copy with me or to provide a link that would enable me to make a copy.  The link I have for the Google form does not provide any menu options.  
Thank you.

**Comment by roslynwythe** [comment]
Hi @macho-catt - is it correct to assume that on the new Homepage Summary, the Display and Homepage columns should be mapped to the corresponding columns in the Review Sheet?  Thank you

**Comment by roslynwythe** [comment]
Progress/Blocker:  I have made code changes on a copy of the wins response sheet in my Google Drive, but to test:
1. I require a copy of the Google wins form https://docs.google.com/forms/d/e/1FAIpQLSeTiFjtxKbYgrJHAdFUovXlE55BdwwUQ9NePOXRY2q-h-w1tg/viewform
2. As described in the Action Items for this issue, for complete testing I need a copy of the Secret files.

**Comment by roslynwythe** [comment]
Progress:
1. The first 4 top-level action items have been completed. 
2. I am currently testing the new code function and sheet formulas in my copy of wins-response.  I also populated the new Homepage Summary sheet with all the recent wins submissions to bring it up to date and in sync with the Review Sheet.  
3. I do have some minor questions which I will pose tonight at the dev office hours.  
4. I expect to be ready to demonstrate the code next week.

**Comment by roslynwythe** [comment]
Progress:
Code and spreadsheet formulas have been written and tested, but I wanted to confirm my approach to managing the common fields between the Homepage Summary and Review sheets.  In particular, is it acceptable to populate these columns on the Homepage Summary sheet using formulas that map to the corresponding columns on the Review sheet:
-  display
- homepage
- member since
- display summary

**Comment by roslynwythe** [comment]
unassigning because dependency https://github.com/hackforla/website/issues/2901 not completed.

**Comment by arpitapandya** [comment]
moving this back to the ice-box as #2901 isn't ready.

**Comment by jdingeman** [comment]
Marking as ready for prioritization

**Comment by roslynwythe** [comment]
- @ExperimentsInHonesty unless this issue is time sensitive, it would be preferable to make it dependent on #3641 and #4134.  Please advise.  By the way, I had virtually completed work on this issue before #2901 was reopened.

**Comment by ExperimentsInHonesty** [comment]
@roslynwythe I moved this into the new issue review column because the dependency is resolved. Please add a ready for prioritization label if its ready.

**Comment by ExperimentsInHonesty** [comment]
@roslynwythe I moved this into the new issue review column because the dependency is resolved. Please add a ready for prioritization label if its ready.

**Comment by roslynwythe** [comment]
@ExperimentsInHonesty questions:
- I assume that the following columns are formulas pulling values from the corresponding columns on the Review sheet:  `Display`, `Homepage`, `Manual Reviewed Role`
- Regarding the columns circled in red below: I assume we are estimating the number of views based on the number of wins entries in the json file.  That calculation will change each time there is a change to the `Homepage` column.  Please let's discuss
<details><summary>Homepage Summary sheet - screenshot</summary>

![image](https://github.com/hackforla/website/assets/5314153/7bf736b4-e34a-4c29-8127-5ef9ea2c3ec2)

</details> 

- can we get started on this now: #4190

**Comment by roslynwythe** [comment]
@ExperimentsInHonesty Looking at "Homepage Summary Prototype" I see that the "Display" and "Homepage" columns are formulas pointing to the Review sheet, but it seems likely that a user might try to change the value of those fields from the "Homepage Summary Prototype" sheet.  How can we prevent that?

**Comment by ExperimentsInHonesty** [comment]
@roslynwythe We could hide it, but that might be a problem later when we need to see it and forget its there.

I just tried changing the color of the font, in case that helps.  But also, I don't see that the Display column is a formula, only the homepage column.  Not sure if that was supposed to connect and never got setup properly.  Should we update it?

>@ExperimentsInHonesty Looking at "Homepage Summary Prototype" I see that the "Display" and "Homepage" columns are formulas pointing to the Review sheet, but it seems likely that a user might try to change the value of those fields from the "Homepage Summary Prototype" sheet. How can we prevent that?

**Comment by roslynwythe** [comment]
@ExperimentsInHonesty Please advise, do you think it makes more sense for PMs to be able to set "Display" and "Homepage" from the Response sheet or from the "Homepage Summary Prototype?   No matter which we choose, I feel like there will be some confusion when PMs are viewing the sheet in which they cannot set those columns.  Perhaps we should combine the "Homepage Summary Protoype" columns into the "Response" sheet?

**Comment by roslynwythe** [comment]
> @roslynwythe We could hide it, but that might be a problem later when we need to see it and forget its there.
> 
> I just tried changing the color of the font, in case that helps. But also, I don't see that the Display column is a formula, only the homepage column. Not sure if that was supposed to connect and never got setup properly. Should we update it?
> 
I updated the `Display` column to be a formula.  The `Homepage` and `Member Since` columns were already formulas.  I also made `Manual Reviewed Role` a formula as well since that seems necessary, and I'm thinking `Display Review` should be a formula as well, although that column could be copied from the Review sheet.  Do you agree?

**Comment by ExperimentsInHonesty** [comment]
@roslynwythe 
- I updated the Review tab to have a drop-down for the role. 
- I don't see where you added a formula for the Manual Reviewed Role on the Homepage summary prototype tab.
- I added to the Review spreadsheet a column for test and populated it with TRUE/FALSE values
- Before we start making more formulas for the content on the homepage summary prototype, let's discuss how we plan to update the columns (what automation and how it will work). The reason I ask this is because, if the columns A-E are all being pulled from Review tab, and then we add new info to Review tab, and then that updates the content in Homepage summary prototype tab, it could potentially disconnect the data in the below columns from the data in A-E.
   - Date Live on homepage start
   - Date Live on homepage end
   - No of times appeared on website

**Comment by roslynwythe** [comment]
> @roslynwythe
> * Before we start making more formulas for the content on the homepage summary prototype, let's discuss how we plan to update the columns (what automation and how it will work). The reason I ask this is because, if the columns A-E are all being pulled from Review tab, and then we add new info to Review tab, and then that updates the content in Homepage summary prototype tab, it could potentially disconnect the data in the below columns from the data in A-E.
>   
>   * Date Live on homepage start
>   * Date Live on homepage end
>   * No of times appeared on website

@ExperimentsInHonesty perhaps we should integrate those homepage summary columns with the Review sheet.

**Comment by ExperimentsInHonesty** [comment]
RW and I met, and we are going to do the following

- [ ] combine the content on the `Homepage summary prototype` tab with the `Review tab` so that there is no risk of the incorrect one being updated.
- [ ] add wins id #2385
- [ ] create a new tab called Log, which should be manually updated every time a new win is manually added to the homepage randomizer.  This way we can see if someone has already been featured and when
- [ ] in the column "# of times appeared on website" it should pull that data from the Log tab by user id
- [ ] add person id 
- [ ] create a new tab called person
- [ ] This should have a list of all people and their ids, and the number of wins
    - schema: id, name, email, # of wins ids

**Comment by roslynwythe** [comment]
@ExperimentsInHonesty the issue has been updated as per your comments in the previous comment.
